src/app/(proper_react)/(redesign)/(authenticated)/user/(dashboard)/settings/CancelFlow.tsx

```diff
@@ -4,42 +4,219 @@
 import { useL10n } from "../../../../../../hooks/l10n";
 import { TelemetryButton } from "../../../../../../components/client/TelemetryButton";
+import { ExperimentData } from "../../../../../../../telemetry/generated/nimbus/experiments";
+import { onApplyCouponCode, onCheckUserHasCurrentCouponSet } from "./actions";
+import { TelemetryLink } from "../../../../../../components/client/TelemetryLink";
 
 export type Props = {
   fxaSubscriptionsUrl: string;
-  confirmationFlagEnabled: boolean;
+  enableDiscountCoupon: boolean;
+  experimentData?: ExperimentData;
+};
+
+type DiscountData = {
+  headline: string;
+  successDescription: string;
+  subtitle: string;
 };
 
 export const CancelFlow = (props: Props) => {
   const l10n = useL10n();
   const recordTelemetry = useTelemetry();
-  const currentStep = props.confirmationFlagEnabled ? "confirm" : "survey";
-  const [step, setCurrentStep] = useState<"confirm" | "survey" | "redirecting">(
-    currentStep,
-  );
+  const [step, setCurrentStep] = useState<
+    "confirm" | "survey" | "all-set" | "redirecting"
+  >("confirm");
 
   const dialogState = useOverlayTriggerState({
     onOpenChange: (isOpen) => {
       recordTelemetry("popup", isOpen ? "view" : "exit", {
         popup_id: "settings-cancel-monitor-plus-dialog",
       });
+      if (isOpen) {
+        setCurrentStep("confirm");
+      }
     },
   });
 
+  const [primaryCta, setPrimaryCta] = useState<ReactElement>();
+  const [ctaSubtitle, setCtaSubtitle] = useState<ReactElement>();
+  const [couponSuccess, setCouponSuccess] = useState<boolean | null>(null);
+  const [alreadyHasCouponSet, setAlreadyHasCouponSet] =
+    useState<boolean>(false);
+
+  const discountedNext3Months: DiscountData = {
+    headline: l10n.getString("settings-unsubscribe-dialog-promotion-cta", {
+      discount_percentage_num: "30%",
+      discount_duration: 3,
+    }),
+    successDescription: l10n.getString(
+      "settings-unsubscribe-dialog-promotion-description",
+      {
+        discount_percentage_num: "30%",
+        discount_duration: 3,
+      },
+    ),
+    subtitle: l10n.getString(
+      "settings-unsubscribe-dialog-promotion-limitations-apply",
+    ),
+  };
+
+  const handleApplyCouponCode = async () => {
+    const result = await onApplyCouponCode();
+    if (result?.success) {
+      setCouponSuccess(true);
+    } else {
+      setCouponSuccess(false);
+    }
+  };
+
+  const checkCouponCode = async () => {
+    const result = await onCheckUserHasCurrentCouponSet();
+    if (typeof result.success === "boolean") {
+      setAlreadyHasCouponSet(result.success);
+    } else {
+      setAlreadyHasCouponSet(false);
+    }
+  };
+
+  useEffect(() => {
+    void checkCouponCode();
+  }, []);
+
+  useEffect(() => {
+    if (couponSuccess) {
+      setCurrentStep("all-set");
+    }
+  }, [couponSuccess]);
+
+  useEffect(() => {
+    // Current experiment
+    // Note: Opted to use state here to support future experiments
```

Comment:
As mentioned above, I don’t think you would need to store the elements in state.